esp-config/src/bin/esp-config/main.rs

```diff
@@ -0,0 +1,297 @@
+        }
+
+        if let Some(errors) = check_after_changes(&work_dir, args.chip, args.config_file.clone())? {
+            errors_to_show = Some(errors);
+        } else {
+            println!("Updated configuration...");
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+fn apply_config(
+    path: &Path,
+    updated_cfg: Vec<CrateConfig>,
+    previous_cfg: Vec<CrateConfig>,
+    cfg_file: Option<String>,
+) -> Result<(), Box<dyn Error>> {
+    let config_toml_path = path.join(cfg_file.as_deref().unwrap_or(DEFAULT_CONFIG_PATH));
+
+    let mut config = std::fs::read_to_string(&config_toml_path)?
+        .as_str()
+        .parse::<DocumentMut>()?;
+
+    if !config.contains_key("env") {
+        config.insert("env", Item::Table(Table::new()));
+    }
+
+    let envs = config.get_mut("env").unwrap().as_table_mut().unwrap();
+
+    for cfg in updated_cfg {
+        let previous_crate_cfg = previous_cfg.iter().find(|c| c.name == cfg.name);
+
+        for option in cfg.options {
+            let previous_option = previous_crate_cfg.and_then(|c| {
+                c.options
+                    .iter()
+                    .find(|o| o.option.name == option.option.name)
+            });
+            let key = option.option.full_env_var(&cfg.name);
+
+            // avoid updating unchanged options to keep the comments (if any)
+            if Some(&option.actual_value) != previous_option.map(|option| &option.actual_value) {
+                if option.actual_value != option.option.default_value {
+                    let value = toml_edit::Value::String(Formatted::new(format!(
+                        "{}",
+                        option.actual_value
+                    )));
+
+                    envs.insert(&key, Item::Value(value));
+                } else {
+                    envs.remove(&key);
+                }
+            }
+        }
+    }
+
+    std::fs::write(&config_toml_path, config.to_string().as_bytes())?;
+
+    Ok(())
+}
+
+fn parse_configs(
+    path: &Path,
+    chip_from_args: Option<esp_metadata::Chip>,
+    config_file: Option<String>,
+) -> Result<Vec<CrateConfig>, Box<dyn Error>> {
+    let config_toml_path = path.join(config_file.as_deref().unwrap_or(DEFAULT_CONFIG_PATH));
+    let config_toml_content = std::fs::read_to_string(config_toml_path)?;
+    let config_toml = config_toml_content.as_str().parse::<DocumentMut>()?;
+
+    let envs: HashMap<String, String> = config_toml
+        .get("env")
+        .and_then(Item::as_table)
+        .map(|table| {
+            table
+                .iter()
+                .filter_map(|(k, v)| v.as_str().map(|s| (k.to_string(), s.to_string())))
+                .collect()
+        })
+        .unwrap_or_default();
+
+    let meta = cargo_metadata::MetadataCommand::new()
+        .current_dir(path)
+        .exec()?;
+
+    // try to guess the chip from the metadata by looking at an active chip feature
+    // for esp-hal
+    let chip_from_meta = {
```

Comment:
If we make these closures, they'll be executed lazily below, instead of always as it is right now.